File: 3267-count-almost-equal-pairs-ii/3267-count-almost-equal-pairs-ii.py
class Solution:
    def countPairs(self, nums: List[int]) -> int:
        dic = defaultdict(int)
        ans = 0
        nums.sort(reverse = True)
        for x in nums:
            ans += dic[x]
            st = set()
            st.add(x)
            n = len(str(x))
            for i in range(n):
                for j in range(i, n):
                    for k in range(n):
                        for l in range(k, n):
                            s = list(str(x))
                            s[i], s[j] = s[j], s[i]
                            s[k], s[l] = s[l], s[k]
                            st.add(int("".join(s)))
            for u in st:
                dic[u] += 1
        return ans

File: 3267-count-almost-equal-pairs-ii/3267-count-almost-equal-pairs-ii.cpp
class Solution {
public:
    const int L=7;
    int countPairs(vector<int>& nums) {
        int n=nums.size();
        unordered_map<string,int> mp,id;
        int ans=0;
        for(int i=1;i<=n;i++){
            int x=nums[i-1];
            string s;
            for(int j=0;j<L;j++){
                s+=(x%10)+'0';
                x/=10;
            }
            ans+=mp[s];
            mp[s]++;
            id[s]=i;
            for(int p=0;p<L;p++){
                for(int q=p+1;q<L;q++){
                    if(s[p]==s[q]) continue;
                    swap(s[p],s[q]);
                    if(id[s]<i){
                        id[s]=i;
                        mp[s]++;
                    }
                    for(int r=p;r<L;r++){
                        for(int t=r+1;t<L;t++){
                            if(s[r]==s[t]) continue;
                            swap(s[r],s[t]);
                            if(id[s]<i){
                                id[s]=i;
                                mp[s]++;
                            }
                            swap(s[r],s[t]);
                        }
                    }
                    swap(s[p],s[q]);
                }
            }
        }
        return ans;
    }
};
